### DISCUSSION TITLE
take_sample function error.

### DISCUSSION THREAD

**Discussion by isarib01** [zubkov-andrei/pg_profile#73] [General]
I am seeing an error when I run the function take_sample to a remote DB.   I can confirm the DB that has the pg_profile extension is running from is completely empty, so I am unsure where the duplicate key violation error is from?

  `                                         List of installed extensions
        Name        | Version |   Schema   |                              Description
--------------------+---------+------------+------------------------------------------------------------------------
 dblink             | 1.2     | profile    | connect to other PostgreSQL databases from within a database
 pg_profile         | 4.2     | profile    | PostgreSQL load profile repository and report builder
 pg_stat_statements | 1.10    | profile    | track planning and execution statistics of all SQL statements executed
 plpgsql            | 1.0     | pg_catalog | PL/pgSQL procedural language
`
FYI, Remote DB has pg_stat_statements installed, and loaded.

`                                                               take_sample
------------------------------------------------------------------------------------------------------------------------------------------
 ("duplicate key value violates unique constraint ""pk_last_stat_satements_srv2""                                                        +
 SQL statement ""INSERT INTO last_stat_statements (                                                                                      +
         server_id,                                                                                                                      +
         sample_id,                                                                                                                      +
         userid,                                                                                                                         +
         username,                                                                                                                       +
         datid,                                                                                                                          +
         queryid,                                                                                                                        +
         plans,                                                                                                                          +
         total_plan_time,                                                                                                                +
         min_plan_time,                                                                                                                  +
         max_plan_time,                                                                                                                  +
         mean_plan_time,                                                                                                                 +
...
         temp_blk_read_time  double precision,                                                                                           +
         temp_blk_write_time double precision                                                                                            +
       )""                                                                                                                               +
 PL/pgSQL function collect_pg_stat_statements_stats(jsonb,integer,integer,integer) line 256 at SQL statement                             +
 SQL statement ""SELECT collect_pg_stat_statements_stats(server_properties, sserver_id, s_id, topn)""                                    +
 PL/pgSQL function take_sample(integer,boolean) line 596 at PERFORM                                                                      +
 PL/pgSQL function take_sample(name,boolean) line 16 at assignment                                                                       +
 Key (server_id, sample_id, userid, datid, queryid, toplevel)=(2, 1, 16389, 16401, -3470897278329642964, t) already exists.",00:00:01.98)
(1 row)
`

**Comment by zubkov-andrei** [comment]
Hello!
Looks like a bug. What is your postgres version?

**Reply by isarib01** [reply]
PG14 on the remote DB's, while PG15 on the local where pg_profile is running on.

Edited. Tried pg_profile on PG14 today, same issue presented.

**Reply by zubkov-andrei** [reply]
Can you check the pg_stat_statements view for duplicates on fields shown in error (userid, datid, queryid, toplevel)? I don't think you'll find one, but it is better to check...

**Reply by isarib01** [reply]
Just checked the remote DB pg_stat_statements view, and there appears to be an oddity duplicate record. Not sure how this is possible by pg_stat_statements. Unless pg_profile can merge duplicate behaviours when aggregating from the pg_stat_statements views.

Edit:
![image](https://github.com/zubkov-andrei/pg_profile/assets/73514261/10d622cd-e308-4118-9634-b1d0812e7459)

**Reply by zubkov-andrei** [reply]
Thank you for checking. In screenshot provided I can't see the toplevel field. Can you please make sure that the value of toplevel field is also equal for those entries? pg_stat_statements should use a hashtable on four fields for tracking statistics, those are userid, datid, queryid, toplevel. We should check them all to make sure.
If you'll see a complete duplicates (I still don't think it is possible in a hashtable), please check some statistic values also. Thank for your research!

**Reply by isarib01** [reply]
Just updated the image.   There does not appear to be a toplevel fields in pg_stat_statements 1.5

![image](https://github.com/zubkov-andrei/pg_profile/assets/73514261/cf6ddb6a-404a-4472-be20-f70f7fd4bce5)

**Reply by zubkov-andrei** [reply]
Hm, you've mentioned PG versions 14 and 15. They should have pg_stat_statements extension version 1.9 and 1.10... Anyway, in your case there are some duplicates exists with different query texts.
Right now we should decide should it be fixed in pg_stat_statements (if it is pgss bug), or pg_profile should be ready for such cases. Please check are those entries have the same statistics values. Maybe this is really the one entry with the same stats duplicated in output someway..

**Comment by panatamann** [comment]
Catched same situation some time ago after upgrade pg12 -> pg14. Root cause was in different versions of pg_stat_statements lib and objects. In pg14 pgss library became 1.9 but extension version remained 1.5. _alter extension pg_stat_statements update_ resolved the problem.

**Reply by isarib01** [reply]
Great!, Will check out what is required to get it upgraded.  Seems like the version has fallen behind since an upgrade.

**Reply by zubkov-andrei** [reply]
@isarib01, if provided solution will fix your problem, let us know. I think it should be marked as answer and this discussion can be closed.

**Reply by isarib01** [reply]
Will mark as answered for now - Need to do some more testing with the suggestions here. cheers.
